raft/basic_raft_fixture_test: deflake

joe-redpanda opened the pull request:
test_delayed_snapshot_request previously flaked because a cluster of a bug in how it performed reconfiguration. the raft group would reconfigure from 3 to 1 node, and then back from 1 to 3, but this test used revision_id = 0 for all stages of this process.

This caused a race on in_flight_appends within follower metrics, because if consensus uses an raii guard to increment and decrement in_flight_appends to suppress heartbeats. If a guard opened, the group reconfigured down, and then the group reconfigured up, and then the guard expired, the guard would attempt to decrement a freshly created follower metrics object, which would attempt to decrement the counter below 0.

The fix is to reconfigure from 1 -> 3 with new node revision ids. This requires shutting down the old nodes after reconfigure down, and starting new nodes before reconfiguring back up. Because follower metrics are tracked by vnode, lingering guards will no longer inappropriately decrement the new metrics.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

bharathv commented:
Fix in hb manager and the test changes are unrelated, right? Perhaps split into two PRs since the the fix has to be backported to all branches?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c5d51bfb-5aac-4320-a18f-714a621e043a-->
<!--TEST RESULTS-->
<!--04a91ac2-3bc8-428c-bf67-0089717a3f84-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69648">build#69648</a></summary>


| test_class        | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| STSRoleFetchTests | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69648#01983efd-6701-40ab-8ad7-86b1afa64bae | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69696">build#69696</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69696#019843c9-c562-41e4-9582-f09303c657cc | FLAKY         | 20/21    | upstream reliability is '99.53917050691244'. current run reliability is '95.23809523809523'. drift is 4.30108 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -76,80 +75,94 @@ heartbeat_manager::requests_for_range() {
+            const auto raft_metadata = r->meta();
+
+            // lightweight heartbeat optimization, requires less data
+            if (should_lw_heartbeat(
+                  r, follower_metadata, raft_metadata, r->flushed_offset())) {
+                r->_probe->lw_heartbeat();
+                // we do not fill the dirty offset and follower request
+                // sequence here as those fields are not used to process
+                // lightweight heartbeats
+                hearbeat_queue.emplace_back(
+                  group_beat,
+                  heartbeat_manager::follower_request_meta(
+                    r, raft::follower_req_seq{}, model::offset{}, id));
+                continue;
+            }
+
+            // full heartbeat path
+            vlog(r->_ctxlog.trace, "[{}] full heartbeat", id);
+            r->_probe->full_heartbeat();
+            const auto seq_id = follower_metadata.next_follower_sequence();
+
+            follower_metadata.last_sent_protocol_meta = raft_metadata;
+            group_beat.data = heartbeat_request_data{
+              .source_revision = r->_self.revision(),
+              .target_revision = id.revision(),
+              .commit_index = raft_metadata.commit_index,
+              .term = raft_metadata.term,
+              .prev_log_index = raft_metadata.prev_log_index,
+              .prev_log_term = raft_metadata.prev_log_term,
+              .last_visible_index = raft_metadata.last_visible_index,
+            };
+            hearbeat_queue.emplace_back(
+              group_beat,
+              heartbeat_manager::follower_request_meta(
+                r, seq_id, raft_metadata.prev_log_index, id));
+
+            if (r->should_reconnect_follower(follower_metadata)) {
+                reconnect_nodes.insert(id.id());
+            }
+        }
```
We should not do it as it may lead to reactor stalls. It would be better to detect mutation of the follower metadata and make sure we are not mutating the entry if it doesn't match after mutation

joe-redpanda replied:
I could reasonably expect many consensus groups on a given shard but not that that many nodes within those, no?

It might be more reasonable, then, to do the async fan-out on the consensus side of things.

```cpp
for (auto& r : _consensus_groups) {
        if (!r->is_elected_leader()) {
            continue;
        }

        co_await ssx::async_for_each_counter(
          counter,
          r->_fstats.begin(),
          r->_fstats.end(),
          [this, r, last_heartbeat, &pending_beats, &reconnect_nodes]...
```


could instead be swapped for something like
```cpp
// snap a view of the raft groups s.t. they don't change under our feet
consensus_set snapped_raft_groups = _consensus_groups;

co_await ssx::async_for_each(snapped_raft_groups.begin(), snapped_groups.end(), [this]() {
    // the snapped sptr garauntees we won't dangling reference a raft group
    // from here, do sync execution s.t. our usage of follower metadata is atomic
     for(auto const& follower_metadata : raft->_fstats) {... sync execution
```

joe-redpanda replied:
or if we're not a fan of snapping, we could just mutex the modification of raft groups s.t. the creation of hb requests can use it un-mutated

joe-redpanda replied:
https://github.com/redpanda-data/redpanda/pull/26996
punted the hb changes here

mmaslankaprv commented:
> Fix in hb manager and the test changes are unrelated, right? Perhaps split into two PRs since the the fix has to be backported to all branches?

+1

joe-redpanda committed:
raft/basic_raft_fixture_test: deflake & hb fix

test_delayed_snapshot_request previously flaked because a cluster of a
bug in how it performed reconfiguration. the raft group would
reconfigure from 3 to 1 node, and then back from 1 to 3, but this test
used revision_id = 0 for all stages of this process.

This caused a race on in_flight_appends within follower metrics, because
if consensus uses an raii guard to increment and decrement
in_flight_appends to suppress heartbeats. If a guard opened, the group
reconfigured down, and then the group reconfigured up, and then the
guard expired, the guard would attempt to decrement a freshly created
follower metrics object, which would attempt to decrement the counter
below 0.

BenPope commented:
Any reason this isn't also flaky in v25.2.x?

joe-redpanda commented:
> Any reason this isn't also flaky in v25.2.x?

Good point, I'll backport

joe-redpanda commented:
/backport v25.1.x v24.3.x v25.2.x

joe-redpanda commented:
/backport v24.3.x

joe-redpanda commented:
/backport v25.2.x
